### PR TITLE
[3.5] Feature/decoupled io

### PR DESCRIPTION
**Pull request by lamai93**
**This is a PR for 3.5.**

Decouples Io from the Scheduler. It introduces multiple `io_contexts` each having one thread. Clients and Endpoints are distributed across the contexts. The new option `--server.io-contexts` can be used to adjust the number of contexts/threads. It defaults to 8. IO threads have the name `Io`.

IO Context Threads handle read/write to sockets, `HTTPCommTask`s and `VSTCommTask`s. The parsed request is then given to the Scheduler.

Fixes arangodb/team-clifton-planning#26

### CODE REVIEW

File: arangod/Scheduler/Socket.h [hunk 1]
```diff
@@ -111,13 +112,12 @@ class Socket {
   virtual void shutdownSend(asio_ns::error_code& ec) = 0;
 
  protected:
-  // strand to ensure the connection's handlers are not called concurrently.
-  std::unique_ptr<asio_ns::io_context::strand> _strand;
```
[the remark this answers is not in the document]
Comment (lamai93) [reply arangod/Scheduler/Socket.h]: Yes, as written above, each `io_context` is managed by a single thread.

[the remark this answers is not in the document]
Comment (mpoeter) [reply arangod/Scheduler/Socket.h]: Ah, right. And every task is bound to a specific io_context therefore ensuring the required serialization. Nice!

**Comment by lamai93** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/999/

File: arangod/GeneralServer/GeneralServer.cpp [hunk 2]
```diff
@@ -40,16 +40,25 @@ using namespace arangodb::rest;
...
 
 void GeneralServer::startListening() {
+  unsigned int i = 0;
+
   for (auto& it : _endpointList->allEndpoints()) {
     LOG_TOPIC(TRACE, arangodb::Logger::FIXME) << "trying to bind to endpoint '"
                                               << it.first << "' for requests";
 
-    bool ok = openEndpoint(it.second);
+    // distribute endpoints acorss all io contexts
```
Comment (neunhoef) [inline arangod/GeneralServer/GeneralServer.cpp]: Typo in comment.

File: arangod/GeneralServer/GeneralServerFeature.cpp [hunk 3]
```diff
@@ -109,12 +109,14 @@ GeneralServerFeature::GeneralServerFeature(
 )
     : ApplicationFeature(server, "GeneralServer"),
       _allowMethodOverride(false),
-      _proxyCheck(true) {
+      _proxyCheck(true),
+      _numIoThreads(8) {
```
Comment (neunhoef) [inline arangod/GeneralServer/GeneralServerFeature.cpp]: I wonder whether this is the right default. Maybe something depending on the number of cores available might be better.

File: arangod/Scheduler/Socket.h [hunk 4]
```diff
@@ -85,10 +86,10 @@ class Socket {
   }
 
   void post(std::function<void()> handler) {
-    _scheduler->post(*_strand, handler);
+    _context.post(handler);
   }
 
-  bool runningInThisThread() { return _strand->running_in_this_thread(); }
+  bool runningInThisThread() { return true; }
```
Comment (neunhoef) [inline arangod/Scheduler/Socket.h]: This rings a (small) alarm bell. We need to check that we do not actually run any request in the io thread and places where this is called might hint us as to where this is happening.

**Comment by Simran-B** [comment]
FYI: The startup option page for category "server" just moved in the docs: `Manual/Programs/Arangod/Server.md`

File: arangod/Scheduler/Socket.h (hunk 4 above)
Comment (lamai93) [reply arangod/Scheduler/Socket.h]: Agree. I will create a `runningInThisThread` functionality for the `IoContext`. In addition to that there was only one place where it was not used in a assertion. However, due to other conditions evaluating always true, a potentially problematic code path was never taken.

**Comment by lamai93** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/1002/

**Comment by lamai93** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/1211/

**Comment by lamai93** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/1410/

**Pull request merged**